pkg/sql/catalog/tabledesc/validate_version_gating_test.go

```diff
@@ -104,6 +105,10 @@ func TestIndexDoesNotStorePrimaryKeyColumnMixedVersion(t *testing.T) {
 	// Deserialize and insert the corrupted descriptor to cluster.
 	const corruptDescToInject = "0ade040a01741868206428133a0042260a016910011a0c0801104018003000501460002000300068007000780080010088010098010042260a016a10021a0c0801104018003000501460002000300068007000780080010088010098010042260a016b10031a0c080110401800300050146000200030006800700078008001008801009801004804526b0a06745f706b65791008180122016a2a01692a016b300240004a10080010001a00200028003000380040005a00700170037a0408002000800100880101900104980101a20106080012001800a80100b20100ba0100c00100c80100d00106e00100e90100000000000000005a6b0a036964781004180022016b2a016a300340004a10080010001a00200028003000380040005a0070027a0408002000800100880101900104980100a20106080012001800a80100b20100ba0100c00100c801f0bc93e0a5c788d017d00102e00100e90100000000000000005a6c0a07745f695f6b6579100618012201693001380240004a10080010001a00200028003000380040005a007a0408002000800100880101900104980100a20106080012001800a80100b20100ba0100c00100c801d8b39aa5a6c788d017d00104e00100e9010000000000000000600a6a210a0b0a0561646d696e100218020a0a0a04726f6f74100218021204726f6f741803800101880103980100b2011c0a077072696d61727910001a01691a016a1a016b2001200220032800b80101c20100e80100f2010408001200f801008002009202009a020a08d0d0d0febfc688d017b20200b80200c00265c80200e00200800300880308a80300b00300d00300"
 	mustInsertDescToDB(ctx, t, sqlDB, corruptDescToInject)
+	// Upgrade precondition checks for corrupt descriptor AOST '-10s', so, we wait
+	// for 10s here to ensure the corruption will be seen by the time the upgrade
+	// precondition check runs.
+	time.Sleep(10 * time.Second)
```

Comment:
i am not a fan of adding such a long sleep in tests. have you considered adding a testing hook like this?
```
diff --git a/pkg/upgrade/upgrades/first_upgrade.go b/pkg/upgrade/upgrades/first_upgrade.go
index 1505eb81702..813718b93d4 100644
--- a/pkg/upgrade/upgrades/first_upgrade.go
+++ b/pkg/upgrade/upgrades/first_upgrade.go
@@ -97,6 +97,14 @@ func upgradeDescriptors(
 	})
 }
 
+var firstUpgradeAOST = true
+
+func TestingSetFirstUpgradeAOST(enabled bool) func() {
+	old := firstUpgradeAOST
+	firstUpgradeAOST = enabled
+	return func() { firstUpgradeAOST = old }
+}
+
 // FirstUpgradeFromReleasePrecondition is the precondition check for upgrading
 // from any supported major release.
 //
@@ -118,7 +126,7 @@ func FirstUpgradeFromReleasePrecondition(
 	// a diagnostic query. If no corruptions were found back then, we assume that
 	// there are no corruptions now. Otherwise, we retry and do everything
 	// without an AOST clause henceforth.
-	withAOST := true
+	withAOST := firstUpgradeAOST
 	diagnose := func(tbl string) (hasRows bool, err error) {
 		q := fmt.Sprintf("SELECT count(*) FROM \"\".crdb_internal.%s", tbl)
 		if withAOST {
```

Reply:
This is a good idea! Done